### PR TITLE
Postgres zero-downtime migrations

### PR DESCRIPTION
**Pull request by oetherington**
This PR implements zero-downtime migrations for Postgres. I'm also particularly interested in any ideas as to improving the process or making it more secure. Pretty much all of the code in this PR can be deleted once the migration process is complete.

The precedure for migrating a collection is as follows:

- Deploy a PR that sets `collectionType: 'switching'` on all the desired collections.
- Run `./scripts/serverShellCommand.sh --wait "Vulcan.migrateCollections(['Posts', 'PostRels'])"` (or whatever collections are being migrated).
- All changes are done in a transaction, so everything will be rolled back if an error occurs.
- If everything completes successfully, the server will now automatically be using Postgres.
- Deploy another PR permanently setting `collectionType: 'pg'`.

This PR also introduces a lock table in Postgres to keep track of which collections have been successfully migrated. This serves two purposes:
- If a server instance gets restarted for some reason after migrating but before deploying the second PR, it will read the value from the lock table and use Postgres instead of Mongo for the relevant collections.
- To handle the fact that we have multiple server instances that need to switch at the ~same time, each collection with `collectionType: 'switching'` polls the lock table once every five seconds until the lock table is set to Postgres, then it stops polling after switching.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203337486019507) by [Unito](https://www.unito.io)
┆Link To Task: https://app.asana.com/0/1201302964208280/1203337486019507

### CODE REVIEW

File: packages/lesswrong/server/scripts/migrateCollections.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+type Transaction = ITask<{}>;
+
+// Custom formatters to fix data integrity issues on a per-collection basis
+// A place for nasty hacks to live...
+const formatters: Partial<Record<CollectionNameString, (document: DbObject) => DbObject>> = {
+  Posts: (post: DbPost): DbPost => {
+    const scoreThresholds = [2, 30, 45, 75, 125, 200];
+    for (const threshold of scoreThresholds) {
+      const prop = `scoreExceeded${threshold}Date`;
```
Comment (Will-Howard) [inline packages/lesswrong/server/scripts/migrateCollections.ts:20]: Why is this necessary? Are these not normal fields?

File: packages/lesswrong/server/scripts/migrateCollections.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+      const copyStart2 = new Date();
+      await copyData(transaction, collections, 2, copyStart);
+
+      // Start writing to Postgres
+      for (const collection of collections) {
+        collection.setTargets("mongo", "pg");
+      }
+
+      // Once more for luck...
+      await copyData(transaction, collections, 3, copyStart2);
```
Comment (Will-Howard) [inline packages/lesswrong/server/scripts/migrateCollections.ts:162]: nice

Comment (Will-Howard) [reply packages/lesswrong/server/scripts/migrateCollections.ts:162]: Although won't the other servers still be writing to mongo at this point so some stuff could end up not being copied over? I think the process should be:
 - first two data copies
 - disable writes to mongo (for all servers, wait for some signal via the lock table that all servers have done this)
 - third data copy
 - enable reads and writes to postgres (for all servers, again wait for signal via the lock table that all servers have done this successfully)

I guess having everything in a transaction makes it hard to communicate via the lock table, could the lock table be a mongo collection instead?

File: packages/lesswrong/server/scripts/migrateCollections.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+      console.log("\ud83c\udfc1 Done!");
+    });
+  } catch (e) {
+    for (const collection of collections) {
+      collection.setTargets("mongo", "mongo");
+    }
+    console.error("Error:", e.message);
+    console.log("\ud83d\uded1 Migration aborted!");
+  }
```
Comment (Will-Howard) [inline packages/lesswrong/server/scripts/migrateCollections.ts:178]: I think it would be good to have the ability to manually switch back to mongo quickly after the fact, e.g. if we see that 5xx errors go way up we should be able to revert it in under 30 seconds to minimise the amount of data loss. Maybe this should even be automatic although that would just introduce a new risk

File: packages/lesswrong/lib/SwitchingCollection.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
+import { MongoCollection } from "./mongoCollection";
+import PgCollection from "./sql/PgCollection";
+import { getCollectionLockType, setCollectionLockType } from "./mongo2PgLock";
+
+export type ReadTarget = "mongo" | "pg";
+export type WriteTarget = ReadTarget | "both";
+
+const POLL_RATE_SECONDS = 5;
```
Comment (Will-Howard) [inline packages/lesswrong/lib/SwitchingCollection.ts:8]: Relevant to my comments below about the possibility of data not being copied over, we may as well make this shorter if there's really very little overhead. We only have about 15 req/s so making it 1 or 2 seconds gets the potential for getting into a weird state pretty low

File: packages/lesswrong/server/scripts/migrateCollections.ts (hunk 1 above)
Comment (oetherington) [reply packages/lesswrong/server/scripts/migrateCollections.ts:20]: This is just cleaning up garbage values in that exist in the database due to old bugs and that violate the Postgres schema. As we migrate more collections, we'll probably find more data integrity issues like this (and Jim thinks LessWrong will have even more due to the size and age of their database).

File: packages/lesswrong/server/scripts/migrateCollections.ts (hunk 3 above)
Comment (oetherington) [reply packages/lesswrong/server/scripts/migrateCollections.ts:178]: That's a good point - this should be solvable by just continuing to poll even after we switch to postgres.

File: packages/lesswrong/lib/SwitchingCollection.ts (hunk 4 above)
Comment (oetherington) [reply packages/lesswrong/lib/SwitchingCollection.ts:8]: Fair point - I'll set it to 1 second for now which should be fine. Even 0.5 seconds might be ok given that the db and the server are in the same data centre 🤔.

File: packages/lesswrong/server/scripts/migrateCollections.ts (hunk 2 above)
Comment (oetherington) [reply packages/lesswrong/server/scripts/migrateCollections.ts:162]: Ah, yeah - you're right. That _is_ the process that we're using right now (when we first set the write target to "pg", that implicitly disables mongo writes), but, as you point out, we only do that on the machine doing the copying, not the server instances, so we need to propogate that through the lock table.

We can still execute SQL queries here without worrying about the transaction - we just have to use `sql` instead of `transaction`.

The main question is, is it sufficient to just write to the lock table and then wait a certain period of time (3 * POLL_RATE_SECONDS maybe?) and trust that the other instances switched, or do we really need two-way communication through the lock? If we do need two-way comms then that's a lot more complicated - I think the simplest way is to have some counter that we can atomically increment, but then we still need to hard code the number of instances which is annoying, and I'm pretty sure it's different for LessWrong.

**Comment by oetherington** [comment]
I'm going to merge this now since it's not going to be turned on quite yet, then if we need any more changes we can make a separate PR.
